pkg/license/aggregator.go

```diff
@@ -0,0 +1,216 @@
+type Aggregator struct {
+	client k8s.Client
+}
+
+type aggregate func() (resource.Quantity, error)
+
+// AggregateMemory aggregates the total memory of all Elastic managed components
+func (a Aggregator) AggregateMemory() (resource.Quantity, error) {
+	var totalMemory resource.Quantity
+
+	for _, f := range []aggregate{
+		a.aggregateElasticsearchMemory,
+		a.aggregateKibanaMemory,
+		a.aggregateApmServerMemory,
+	} {
+		memory, err := f()
+		if err != nil {
+			return resource.Quantity{}, err
+		}
+		totalMemory.Add(memory)
+	}
+
+	return totalMemory, nil
+}
+
+func (a Aggregator) aggregateElasticsearchMemory() (resource.Quantity, error) {
+	var esList esv1.ElasticsearchList
+	err := a.client.List(&esList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, es := range esList.Items {
+		for _, nodeSet := range es.Spec.NodeSets {
+			mem, err := containerMemLimits(
+				nodeSet.PodTemplate.Spec.Containers,
+				esv1.ElasticsearchContainerName,
+				essettings.EnvEsJavaOpts,
+				memFromJavaOpts,
+				nodespec.DefaultMemoryLimits,
+			)
+			if err != nil {
+				return resource.Quantity{}, err
+			}
+
+			total.Add(multiply(mem, nodeSet.Count))
+			log.V(1).Info("Collecting", "es_name", es.Name, "memory", mem.String(), "count", nodeSet.Count)
+		}
+	}
+
+	return total, nil
+}
+
+func (a Aggregator) aggregateKibanaMemory() (resource.Quantity, error) {
+	var kbList kbv1.KibanaList
+	err := a.client.List(&kbList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, kb := range kbList.Items {
+		mem, err := containerMemLimits(
+			kb.Spec.PodTemplate.Spec.Containers,
+			kbv1.KibanaContainerName,
+			kbconfig.EnvNodeOpts,
+			memFromNodeOptions,
+			kbpod.DefaultMemoryLimits,
+		)
+		if err != nil {
+			return resource.Quantity{}, err
+		}
+
+		total.Add(multiply(mem, kb.Spec.Count))
+		log.V(1).Info("Collecting", "kibana_name", kb.Name, "memory", mem.String(), "count", kb.Spec.Count)
+	}
+
+	return total, nil
+}
+
+func (a Aggregator) aggregateApmServerMemory() (resource.Quantity, error) {
+	var asList apmv1.ApmServerList
+	err := a.client.List(&asList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, as := range asList.Items {
+		mem, err := containerMemLimits(
+			as.Spec.PodTemplate.Spec.Containers,
+			apmv1.ApmServerContainerName,
+			"",
+			nil,
+			apmserver.DefaultMemoryLimits,
+		)
+		if err != nil {
+			return resource.Quantity{}, err
+		}
+
+		total.Add(multiply(mem, as.Spec.Count))
+		log.V(1).Info("Collecting", "as_name", as.Name, "memory", mem.String(), "count", as.Spec.Count)
+	}
+
+	return total, nil
+}
+
+// containerMemLimits reads the container memory limits from the resource specification with fallback
+// on the environment variable and on the default limits
+func containerMemLimits(
+	containers []corev1.Container,
+	containerName string,
```

Comment:
Would it be simpler if we just pass it the specific container we want to examine rather than the array and the name of the item in the array?

Reply:
Not really.